### PR TITLE
[DBManager]  replace pyspatialite with pysqlite

### PR DESCRIPTION
**Pull request by alexbruy**
Pyspatialite does not support SpatiaLite 4.x. Pysqlite module is much more widely used and now it is possible to load spatial extensions via SQLite's load_extension() mechanism without issues.

See also old discussion in the SpatiaLite mailing list https://groups.google.com/forum/#!topic/spatialite-users/iTOGhiBi83E

### CODE REVIEW

**Comment by luipir** [comment]
in this way would be possibile to have internal pyspatialite compilation?
I suppose that for back compatibility it would be still supported => probably have to do contidianl import. I didn't investigated if in older distribution we have to still have pyspytialite.

**Comment by m-kuhn** [comment]
It is also used in processing/algs/qgis/spatialite_utils.py and it is mentioned in the readme of the db_manager plugin. I assume these two should occurrences should be changed too?

File: python/plugins/db_manager/db_plugins/spatialite/connector.py [hunk 1]
```diff
@@ -44,7 +45,13 @@ def __init__(self, uri):
 
         try:
             self.connection = sqlite.connect(self._connectionInfo())
-
+            self.connection.enable_load_extension(True)
+            c = self._get_cursor()
+            if os.name == 'nt':
```
Comment (m-kuhn) [inline python/plugins/db_manager/db_plugins/spatialite/connector.py]: Is this correct? Is 'nt' not Windows and that's where the .dll should be used?

Would it make sense to put this init code into a helper module (e.g. qgis.spatialite), what do you think?

**Comment by m-kuhn** [comment]
I had trouble with pyspatialite on python 3 so I'm happy if we can use this approach instead.

**Comment by alexbruy** [comment]
Hope now all issues solved. I tested this locally at it works fine for me.

Also I managed to build and package pysqlite for 32-bit OSGeo4W, but stuck with 64-bit. Once packaging issue will be solved this will be completely ready for merge.

**Comment by alexbruy** [comment]
Package for OSGeo4W64 also ready.

**Comment by m-kuhn** [comment]
Great,

can we test linux and osx (with pip install) via travis?

These two tests also use pyspatialite and would be good candidates to test it:

* PyQgsVirtualLayerProvider
* PyQgsSpatialiteProvider

**Comment by alexbruy** [comment]
I tested it locally under Linux and seems all fine. I will look at tests and try to update them

**Comment by alexbruy** [comment]
Tests updated. I also tried to add pysqlite to Travis Linux build, not sure if I did this correctly

**Comment by m-kuhn** [comment]
That may work for the linux/qt4 tests, the others work with pip:

https://github.com/alexbruy/QGIS/blob/fa08d18aeebe9f5bd40fb4cd800fb992d073fa0a/ci/travis/linux/qt5/before_install.sh#L16

https://github.com/alexbruy/QGIS/blob/fa08d18aeebe9f5bd40fb4cd800fb992d073fa0a/ci/travis/linux/qt4/before_install.sh#L9

https://github.com/alexbruy/QGIS/blob/fa08d18aeebe9f5bd40fb4cd800fb992d073fa0a/ci/travis/osx/before_install.sh#L21

And you could remove the tests from the blacklist to check if they now run on python3:

https://github.com/alexbruy/QGIS/blob/fa08d18aeebe9f5bd40fb4cd800fb992d073fa0a/ci/travis/linux/qt5/blacklist.txt#L15-L17

**Comment by alexbruy** [comment]
Linux/Qt4 test failed because pysqlite installation on Travis failed. Also seems there is no pysqlite for Python 3 and we can not use standard sqlite3 module too because it does not support loading extensions.

**Comment by alexbruy** [comment]
Looking if we can use QtSql for this

**Comment by m-kuhn** [comment]
Did you try this instead? It doesn't require pip at all, but I haven't done any further testing apart from the snippet pasted below:

```
$ python3
Python 3.4.3 (default, Mar 31 2016, 20:42:37) 
[GCC 5.3.1 20151207 (Red Hat 5.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlite3 import dbapi2 as sqlite
>>> con = sqlite.connect('')
>>> con.enable_load_extension(True)
>>> cur = con.cursor()
>>> cur.execute("SELECT load_extension('mod_spatialite')")
<sqlite3.Cursor object at 0x7f18b0ca2490>
```

**Comment by jef-n** [comment]
on debian/ubuntu requires ``libsqlite3-mod-spatialite`` available on debian unstable, stretch/testing and ubuntu wily, xenial and yakkety.

**Comment by alexbruy** [comment]
> Did you try this instead?

Under Linux it works, but fails on Windows and maybe also Mac. Because by default extensions loading is disabled, see https://docs.python.org/3.5/library/sqlite3.html#sqlite3.Connection.enable_load_extension

**Comment by m-kuhn** [comment]
We could still use sqlite3 if the QtSql approach fails I guess.

On windows we can ship the module with this option enabled, not sure about mac. On Linux this would mean a minimum requirement for the distributions to use python3 - or backporting the module. And with python2 we can use pysqlite3 for the time being.

Alternative approaches would be porting pyspatialite or pysqlite3 to python 3.

**Comment by mhugo** [comment]
So ... it means we would need a mod_spatialite for windows and mac ?
I've tested to load "spatialite.dll" extension in Windows using standard sqlite3 module and it fails saying it cannot find the right functions in the dll. So it means loading extension is enabled, but that spatialite.dll is not the right module, a mod_spatialite is needed. @alexbruy Am I missing something else ?

**Pull request closed without merging**